std/string.d

```diff
@@ -2962,39 +2962,188 @@ S leftJustify(S)(S s, size_t width, dchar fillChar = ' ') @trusted pure
-        retval[0 .. $ - s.length] = cast(C)fillChar;
-        retval[$ - s.length .. $] = s[];
-        return cast(S)retval;
+    Returns:
+        a lazy range of the right justified result
+
+    See_Also:
+        $(LREF leftJustifier)
+  +/
+
+auto rightJustifier(Range)(Range r, size_t width, dchar fillChar = ' ')
+    if (isForwardRange!Range && isSomeChar!(ElementEncodingType!Range))
+{
+    alias C = Unqual!(ElementEncodingType!Range);
+
+    static if (C.sizeof == 1)
+    {
+        import std.utf : byDchar, byChar;
+        return rightJustifier(r.byDchar, width, fillChar).byChar;
     }
-    else
+    else static if (C.sizeof == 2)
     {
-        auto dstr = to!dstring(s);
-        if (dstr.length >= width)
-            return s;
+        import std.utf : byDchar, byWchar;
+        return rightJustifier(r.byDchar, width, fillChar).byWchar;
+    }
+    else static if (C.sizeof == 4)
+    {
+        static struct Result
+        {
+          private:
+            Range _input;
+            size_t _width;
+            alias _width nfill;       // number of fill characters to prepend
+            dchar _fillChar;
+            bool inited;
+
+            // Lazy initialization so constructor is trivial and cannot fail
+            void initialize()
+            {
+                // Replace _width with nfill
+                // (use alias instead of union because CTFE cannot deal with unions)
+                assert(_width);
+                static if (hasLength!Range)
+                {
+                    auto len = _input.length;
+                    nfill = (_width > len) ? _width - len : 0;
+                }
+                else
+                {
+                    // Lookahead to see now many fill characters are needed
+                    size_t len = 0;
+                    for (auto r = _input.save; true; r.popFront())
+                    {
+                        if (len == _width)
+                        {
+                            nfill = 0;
+                            break;
+                        }
+                        if (r.empty)
+                        {
+                            nfill = _width - len;
+                            break;
+                        }
+                        ++len;
+                    }
+                }
+                inited = true;
+            }
 
-        auto retval = new dchar[](width);
-        retval[0 .. $ - dstr.length] = fillChar;
-        retval[$ - dstr.length .. $] = dstr[];
-        return to!S(retval);
+          public:
+            this(Range input, size_t width, dchar fillChar) pure nothrow
+            {
+                _input = input;
+                _fillChar = fillChar;
+                _width = width;
+            }
+
+            @property bool empty()
+            {
+                return !nfill && _input.empty;
+            }
+
+            @property C front()
+            {
+                C c;
```

Comment:
why use an awkward temporary instead of early returns and simpler control flow?

```d
if (!nfill)
    return _input.front; // fast path
if (!inited)
    initialize();
return nfill ? _fillChar : _input.front;
```